drivers/sms_ser.c

```diff
@@ -0,0 +1,702 @@
+        status_set("BYPASS");
+    } else if (DeviceData.onbattery) {
+        upslogx(LOG_INFO, "setting status to OB");
+        status_set("OB");
+    } else if (DeviceData.lowbattery) {
+        upslogx(LOG_INFO, "setting status to LB");
+        status_set("LB");
+    } else if (!DeviceData.upsok) {
+        upslogx(LOG_INFO, "setting status to RB");
+        status_set("RB");
+    } else if (DeviceData.boost) {
+        upslogx(LOG_INFO, "setting status to BOOST");
+        status_set("BOOST");
+    } else if (!DeviceData.onbattery) {
+        /* sometimes the flag "onacpower" is not set */
+        upslogx(LOG_INFO, "setting status to OL");
+        status_set("OL");
+    } else {
+        /* None of these parameters is ON, but we got some response,
+		 * so the device is (administratively) OFF ? */
+        upslogx(LOG_INFO, "setting status to OFF");
+        status_set("OFF");
+    }
+
+    status_commit();
+    dstate_dataok();
+
+    poll_interval = 5;
+}
+
+void upsdrv_shutdown(void) {
+    /* tell the UPS to shut down, then return - DO NOT SLEEP HERE */
+    int retry;
+
+    /* maybe try to detect the UPS here, but try a shutdown even if
+     * it doesn't respond at first if possible */
+
+    /* replace with a proper shutdown function */
+
+    /* you may have to check the line status since the commands
+     * for toggling power are frequently different for OL vs. OB */
+
+    /* OL: this must power cycle the load if possible */
+
+    /* OB: the load must remain off until the power returns */
+    upsdebugx(2, "upsdrv Shutdown execute");
+
+    for (retry = 1; retry <= MAXTRIES; retry++) {
+        if (sms_instcmd("shutdown.stop", NULL) != STAT_INSTCMD_HANDLED) {
+            continue;
+        }
+
+        if (sms_instcmd("shutdown.return", NULL) != STAT_INSTCMD_HANDLED) {
+            continue;
+        }
+
+        upslogx(LOG_ERR, "Shutting down");
+        set_exit_flag(-2); /* EXIT_SUCCESS */
+        return;
+    }
+
+    upslogx(LOG_ERR, "Shutdown failed!");
+    set_exit_flag(-1);
+}
+
+void upsdrv_help(void) {
+}
+
+/* list flags and values that you want to receive via -x */
+void upsdrv_makevartable(void) {
+    char msg[256];
+
+    upslogx(LOG_INFO, "upsdrv_makevartable");
+
+    snprintf(msg, sizeof msg, "Set reboot delay, in seconds (default=%d).",
+             DEFAULT_BOOTDELAY);
+    addvar(VAR_VALUE, "rebootdelay", msg);
+}
+
+void upsdrv_initups(void) {
+    char *val;
+
+	upslogx(LOG_INFO, "upsdrv_initups");
+
+    upsfd = ser_open(device_path);
+    ser_set_speed(upsfd, device_path, B2400);
+
+    if ((val = getval("rebootdelay"))) {
+        int ipv = atoi(val);
+        if (ipv >= 0)
+            bootdelay = (unsigned int)ipv;
+    }
+}
+
+void upsdrv_cleanup(void) {
+    upslogx(LOG_INFO, "upsdrv_cleanup");
+    /* free(dynamic_mem); */
+    ser_close(upsfd, device_path);
+}
+
+uint8_t sms_prepare_get_status(uint8_t *buffer) {
+    buffer[0] = 'Q';
+    buffer[1] = -1;
```

Comment:
Here and in subsequent methods, for an `uint8_t` array there may be no `-1` value. Maybe better set it as `0xFF` or `255` explicitly?

Revise also the (checksum/CRC?) calculation with `(...) * -1` patterns.

Reply:
the "-1" im just following the protocol, and maybe its OK for me, to set as 0xFF, 255 or NULL. Maybe if change to number or HEX, change all to the same (like change the 'Q' command to a HEX to).

Reply:
Well, to me the actual letters like `'Q'` are okay, human-readable as the actual protocol posted to the device.